torch/csrc/jit/codegen/cuda/index_compute.cpp

```diff
@@ -1940,6 +1880,245 @@ std::pair<std::vector<kir::Val*>, bool> Index::getConsumerRootPredIndices(
+      });
+
+  if (unswitch) {
+    bool within_unswitch = false;
+    const auto one = ir_builder.create<kir::Int>(1);
+    for (auto loop : loops) {
+      if (loop->iter_domain()->parallelType() == ParallelType::Unroll ||
+          loop->iter_domain()->parallelType() == ParallelType::Unswitch ||
+          loop->iter_domain()->parallelType() == ParallelType::Vectorize) {
+        within_unswitch = true;
+      }
+
+      if (within_unswitch) {
+        if (loop->iter_domain()->isBroadcast()) {
+          // Start with a thread binding but still on a broadcast can send
+          // indices through to predicates even if they're not needed below.
+          // Just don't bind anything to the broadcast dim.
+          continue;
+        } else if (loop->iter_domain()->isThread()) {
+          loop_to_ind_map[loop] = loop->start();
+        } else {
+          loop_to_ind_map[loop] = ir_builder.subExpr(loop->stop(), one);
+        }
+      }
+    }
+  }
+
+  std::unordered_map<kir::IterDomain*, kir::Val*> ref_id_to_ind_map;
+  // Due to rfactor/initialization reference_domain may be bigger than loop nest
+  // structure
+  TORCH_INTERNAL_ASSERT(loops.size() <= reference_domain->nDims());
+  for (size_t loop_i = 0; loop_i < loops.size(); loop_i++) {
+    auto ref_axis = gpu_lower->lowerValue(reference_domain->axis(loop_i))
+                        ->as<kir::IterDomain>();
+    ref_id_to_ind_map[ref_axis] = loop_to_ind_map[loops[loop_i]];
+  }
+
+  auto consumer_tv = kir_consumer_tv->fuserTv();
+
+  // Map reference tensor to consumer
+  std::unordered_map<IterDomain*, IterDomain*> root_ref_to_consumer;
+  for (auto c_root : consumer_tv->getMaybeRFactorDomain()) {
+    auto concrete_id = gpu_lower->caIndexMap().getConcreteMappedID(c_root);
+    auto ref_id_it = reference_id_map.find(concrete_id);
+    if (ref_id_it != reference_id_map.end()) {
+      root_ref_to_consumer[ref_id_it->second] = c_root;
+    }
+  }
+
+  BestEffortReplay replay_consumer_as_ref(
+      consumer_tv->domain()->domain(),
+      reference_domain->domain(),
+      root_ref_to_consumer);
+
+  const auto& ref_2_consumer = replay_consumer_as_ref.getReplay();
+
+  // Generate halo information for reference.
+  updateHaloInfoForReference(reference, consumer_tv);
+
+  std::unordered_map<kir::IterDomain*, kir::Val*> reference_halo_extent_map;
+
+  const auto& halo_info = gpu_lower->haloInfo();
+
+  // Generate map from reference iter domains to halo extents
+  for (auto entry : ref_2_consumer) {
+    auto ref_id = entry.first;
+    auto extent = halo_info.getExtent(ref_id);
+    if (extent != nullptr) {
+      reference_halo_extent_map[gpu_lower->lowerValue(ref_id)
+                                    ->as<kir::IterDomain>()] =
+          gpu_lower->lowerValue(extent);
+    }
+  }
+
+  // Index into the reference tensor
+  auto ref_indexing = getReferenceIndexing(
+      loops,
+      reference_domain,
+      ref_id_to_ind_map,
+      {},
+      reference_halo_extent_map);
```

Comment:
While this looks correct to me, I don't think it's actually exercised yet as the shift-specific logic is used when tensors have halo extents. I'm fine having this here, but we may want to leave a note.